openshift/docs/deployment-automation.md

```diff
@@ -0,0 +1,62 @@
+## Github Flow Based Deployments
+
+The deployment pipeline of the microservices in this project are of the Github Flow Type. 
+
+__TL;DR__ 
+- All PRs should be relatively short lived
+- PR code is pushed into a production environment
+- Prior to being merged into master, the source code in master therefor can be defined as always 100% working
+and deployable code. 
+- Each PR will deploy a running instance of the code in the development openshift namespace
+- Each PR's is subsequently promoted to test/prod
+- Not more than 1 PR may be promoted to an environment at the same time
+
+
+## Mantra
+> 'To create the fastest and simplest way to deploy code into production that minimizes toil and jumping through hoops, therefor,
+the utilization of automation will be used at extremes to fast track things that don't or shouldn't require humans to do work'
+
+### Continuos Integration
+
+All continuos integration will be done asynchronously. This means that unit, integration, functional, or lint tests will run and not hinder __deployments to dev__. Each one of these tests then are __a singular github action__
+
+As promotion of code into more sensitive environments occur, deployments into these environments will take pass/failures of CI based github actions into account prior to deploying. 
+
+### Continuos Deployment
+
+> The hope is that a team will have enough skills and the product owner enough confidence to trust that code can go straight into production automatically. This is not something that can be immediately done. A lot of work needs to be put into designing code in a way that allows it to placed into sensitive environments with minimal risk. 
```

Comment:
This still seems like an anti-pattern to me.  Do we still get branch based on our PR to demo to the PO before merging?

Reply:
It really isn't. This is a strategy major companies have adopted, Spotify, Netflix etc all embrace CD. The idea is this. Let your technical team self-organize, give them as much information as you can at the time to get the job done. If things need to demonstration give them that power to demo to stakeholders as needed (hence the Dev PR), reduce the amount of gates that it takes for them to goto prod. To me prod is no different than our 'dev' or 'test' instances. 


By minimizing risks I mean that there are very effective strategies to PR based workflows. For example feature flagging can be implemented, writing good e2e and functional tests to ensure your users are interacting with the app in ways you expect. Infrastructure testing and load testing. 

Again this is a hope, no team in Gov is doing this yet. Typically for a few different reasons:

1. Technical Ability isn't there yet. Not enough confidence in testing or testing technology or automation
2. Leadership endorsement isn't there yet. 
3. Government processes inhibit CD. There's really nothing we can do about this. Some deployments will require hoops to jump in. 


To answer your question yes. Your PR will deploy a instance of the app which you can show to the PO.